### DISCUSSION TITLE
multiple 'src/lib' directories

### DISCUSSION THREAD

**Discussion by mbrightcpacket** [bashly-framework/bashly#256] [Ideas]
The `bashly add lib` functionality is incredibly useful because it prevents code duplication. It would be really nice if we could specify multiple such directories, pulling in all the functions and code defined in them into the final script. In this manner, there could be separate Bashly scripts selecting different combinations of lib directories. For instance, there could be the following three directories:

* aws (AWS related functions)
* azure (Microsoft Azure related functions)
* common (common code that cross-cuts both clouds)

Two bashly scripts could be configured as follows:

* aws.sh could use aws and common
* azure.sh could use azure and common

Is there a way to achieve this in another way?

**Comment by DannyBen** [comment]
Well.

Right now, the source directory for library function is a single directory. There is no support for multiple "load paths".
Note however, that you can have as many subdirectories as you wish inside the `src/lib` directory (so you can group your functions as you please).

Also note that you can specify a different location for that directory in the [`settings.yml`](https://bashly.dannyb.co/usage/settings/) file, or via an environment variable.

But I suspect none of these answer your question directly - I was mainly ensuring we are on the same page.

As I understand, you wish to add another load path for libraries that is external to the directory of the specific bashly project, so that you can use the same lib directory for more than one bashly projects. Correct?

If so, then:

1. It is not supported at this point.
2. It is not a bad idea, I might implement it if there is interest.
3. You might be able to bypass this problem by using a symlink directory inside the lib directory, like this:

```bash
# In the aws project
src/lib/aws     # real directory
src/lib/common  # symlink to /wherever/functions/bashly-common

# In the azure project
src/lib/azure   # real directory
src/lib/common  # symlink to /wherever/functions/bashly-common

```

Not tested, but should work - although I agree, not ideal.

**Comment by mbrightcpacket** [comment]
I was thinking of the symlink approach too, I will go with that for now. Thanks for your quick reply!

**Reply by DannyBen** [reply]
Ok - I will think about it more.
There is high likelihood that the "main" lib dir will have to remain a single one, since it is used by the upgrade function (where you can [upgrade built in libraries](https://bashly.dannyb.co/advanced/library/#auto-update)) - but, it might be relatively easy to add some `also_include` configuration that provides additional directories for inclusion of user directories.

Anyways - let me know how the symlink experiment goes, and if there is a need for a better solution.

**Comment by tzah4748** [comment]
@DannyBen - what was the final result of this? I attempted to use symlink to create a single source of libs for multiple bashly projects but it looks like the symlink is just ignored...

`ls -la ./src/lib` contains `common -> ../../libs`
and `ls -la ../../libs` contains my raw bash scripts.

The scripts under libs worked perfectly fine when put directly under `./src/lib`

**Reply by DannyBen** [reply]
I never tested the symlink myself, it is not something I want to encourage nor support. It is a bad pattern.

But if there is interest, perhaps I can make the `Settings.lib_dir` option accept an array as well.
Note that in any case of multiple lib dirs that I may implement, the "main" lib dir still has a special meaning, and it will be the only one that is upgradeable via `bashly generate --upgrade`.

If this is acceptable, I can try and implement - either as `lib_dir = [lib, ../../external_lib]` or as `additional_lib_dirs = [../../external_lib]`

**Reply by tzah4748** [reply]
Yes please👍🏼
Specifying more than 1 lib folder will result in a much better of adding libraries to multiple different projects

**Comment by DannyBen** [comment]
Sure. I think I am almost done. There will be a new settings option: `extra_lib_dirs: [...]` - which will be relative to the working dir (usually from where you run `bashly g`. I just need to run tests now and implement the new tests for it.

**Reply by DannyBen** [reply]
Implementation in #684

**Reply by tzah4748** [reply]
Awesome, is this already available in bashly:edge ?

**Reply by DannyBen** [reply]
I just merged. It [will be in a few minutes](https://github.com/bashly-framework/bashly/actions/runs/19243705244/job/55012322073).
